engine/table/src/test/java/io/deephaven/engine/table/impl/QueryTableSortedFirstAggTestBase.java

```diff
@@ -0,0 +1,144 @@
+import org.junit.experimental.categories.Category;
+
+import static io.deephaven.engine.testutil.assertj.TableAssert.assertThat;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Category(OutOfBandTest.class)
+public abstract class QueryTableSortedFirstAggTestBase {
+
+    static final String S1 = "S1";
+
+    @Rule
+    public final EngineCleanup base = new EngineCleanup();
+
+    @Before
+    public void setUp() throws Exception {
+        ChunkPoolReleaseTracking.enableStrict();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        ChunkPoolReleaseTracking.checkAndDisable();
+    }
+
+    public abstract Table sortedFirst(char[] source);
+
+    public abstract Table sortedFirst(float[] source);
+
+    public abstract Table sortedFirst(double[] source);
+
+    @Test
+    public void testEmptyChar() {
+        assertThat(sortedFirst(new char[0]).isEmpty()).isTrue();
+    }
+
+    @Test
+    public void testEmptyFloat() {
+        assertThat(sortedFirst(new float[0]).isEmpty()).isTrue();
+    }
+
+    @Test
+    public void testEmptyDouble() {
+        assertThat(sortedFirst(new double[0]).isEmpty()).isTrue();
+    }
+
+    @Test
+    public void testAllNullChar() {
+        check(QueryConstants.NULL_CHAR, new char[] {QueryConstants.NULL_CHAR});
+    }
+
+    @Test
+    public void testAllNullFloat() {
+        check(QueryConstants.NULL_FLOAT, new float[] {QueryConstants.NULL_FLOAT});
+    }
+
+    @Test
+    public void testAllNullDouble() {
+        check(QueryConstants.NULL_DOUBLE, new double[] {QueryConstants.NULL_DOUBLE});
+    }
+
+    @Test
+    public void testAllNaNFloat() {
+        check(Float.NaN, new float[] {Float.NaN});
+    }
+
+    @Test
+    public void testAllNaNDouble() {
+        check(Double.NaN, new double[] {Double.NaN});
+    }
+
+    @Test
+    public void testValuesChar() {
+        check(QueryConstants.NULL_CHAR,
+                new char[] {QueryConstants.NULL_CHAR, Character.MIN_VALUE, Character.MAX_VALUE - 1});
+    }
+
+    @Test
+    public void testValuesFloat() {
+        check(QueryConstants.NULL_FLOAT,
+                new float[] {QueryConstants.NULL_FLOAT, Float.NEGATIVE_INFINITY, Float.POSITIVE_INFINITY, Float.NaN});
+    }
+
+    @Test
+    public void testValuesDouble() {
+        check(QueryConstants.NULL_DOUBLE, new double[] {QueryConstants.NULL_DOUBLE, Double.NEGATIVE_INFINITY,
+                Double.POSITIVE_INFINITY, Double.NaN});
+    }
+
+    @Test
+    public void testNegZeroFirstFloat() {
+        check(-0.0f, new float[] {-0.0f, 0.0f});
+    }
+
+    @Test
+    public void testNegZeroFirstDouble() {
+        check(-0.0, new double[] {-0.0, 0.0});
+    }
+
+    @Test
+    public void testZeroFirstFloat() {
+        check(0.0f, new float[] {0.0f, -0.0f});
+    }
+
+    @Test
+    public void testZeroFirstDouble() {
+        check(0.0, new double[] {0.0, -0.0});
+    }
+
+    public void check(char expected, char[] data) {
+        assertThat(sortedFirst(data))
```

Comment:
TODO: This test should have a sentinel value too, not just the single value under test.